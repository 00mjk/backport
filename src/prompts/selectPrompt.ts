//@ts-ignore
import { MultiSelect, Select, Prompt } from '@sqren/enquirer';
import chalk from 'chalk';
import { HandledError } from '../services/HandledError';

type Choice = {
  name: string;
  displayShort: string;
  displayLong: string;
  enabled: boolean | undefined;
};

export let _currentPrompt: Prompt;

export async function selectPrompt<T extends Choice>({
  message,
  choices,
  isMultiple,
}: {
  message: string;
  choices: T[];
  isMultiple: boolean;
}): Promise<T[]> {
  const promptChoices = [
    ...choices.map((c) => ({
      ...c,
      original: c,
    })),
    { role: 'separator' } as any,
  ];

  type PromptChoice = typeof promptChoices[0];

  // styles for focused item: bold (instead of underlined)
  const styles = {
    em: chalk.cyan.bold,
  };

  if (!isMultiple) {
    const prompt = new Select({
      limit: 20,
      styles,
      message,
      choices: promptChoices,
      format() {
        return chalk.cyan(this.selected.displayLong);
      },
      result() {
        return this.selected.original;
      },
    });

    _currentPrompt = prompt;

    try {
      const answer: T = await prompt.run();
      return [answer];
    } catch (e) {
      if (!e) {
        throw new HandledError('Aborted');
      }
      throw e;
    }
  }

  const prompt = new MultiSelect({
    limit: 20,
    pointer: `❯ `,
    styles,
    message,
    choices: promptChoices,
    initial: promptChoices.filter((c) => c.enabled).map((c) => c.name),
    validate(items: PromptChoice[]) {
      if (items.length === 0) {
        return 'Please use <space> to select at least one option';
      }

      return true;
    },
    format() {
      if (!this.state.submitted) {
        return '';
      }

      // no items selected
      if (this.selected.length === 0) {
        return;
      }

      // 1 item selected
      if (this.selected.length === 1) {
        return chalk.cyan(this.selected[0].displayLong);
      }

      // multiple items selected
      return this.selected
        .map((c: PromptChoice) => chalk.cyan(c.displayShort))
        .join(', ');
    },
    result() {
      return this.selected.map((c: PromptChoice) => c.original);
    },
  });

  _currentPrompt = prompt;

  try {
    const answers = await prompt.run();
    return answers;
  } catch (e) {
    if (!e) {
      throw new HandledError('Aborted');
    }
    throw e;
  }
}
