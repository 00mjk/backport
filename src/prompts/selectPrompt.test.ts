import { selectPrompt, _currentPrompt } from './selectPrompt';

describe('selectPrompt', () => {
  const choices = [
    {
      name: 'black',
      displayLong: 'black as the night',
      displayShort: 'b',
      enabled: false,
    },
    {
      name: 'yellow',
      displayLong: 'yellow as the sun',
      displayShort: 'y',
      enabled: false,
    },
    {
      name: 'red',
      displayLong: 'red as a rose',
      displayShort: 'r',
      enabled: false,
    },
  ];

  it('should return a single answer', async () => {
    const promise = selectPrompt({
      isMultiple: false,
      message: 'Pick ONE color',
      choices,
    });

    const spy = mockPromptWrite();

    _currentPrompt.on('run', async () => {
      await _currentPrompt.keypress(null, { name: 'down' });
      _currentPrompt.submit();
    });

    const answers = await promise;
    expect(answers).toEqual([
      {
        name: 'yellow',
        displayLong: 'yellow as the sun',
        displayShort: 'y',
        enabled: false,
      },
    ]);

    expect(spy.mock.calls).toMatchSnapshot();
  });

  it('should return multiple answers', async () => {
    const promise = selectPrompt({
      isMultiple: true,
      message: 'Pick one or more colors',
      choices,
    });

    const spy = mockPromptWrite();

    _currentPrompt.on('run', async () => {
      await _currentPrompt.keypress(' ');
      await _currentPrompt.keypress(null, { name: 'down' });
      await _currentPrompt.keypress(' ');
      _currentPrompt.submit();
    });

    const answers = await promise;
    expect(answers).toEqual([
      {
        name: 'black',
        displayLong: 'black as the night',
        displayShort: 'b',
        enabled: false,
      },
      {
        name: 'yellow',
        displayLong: 'yellow as the sun',
        displayShort: 'y',
        enabled: false,
      },
    ]);
    expect(spy.mock.calls).toMatchSnapshot();
  });
});

function mockPromptWrite() {
  return jest.spyOn(_currentPrompt, 'write').mockImplementation(() => '');
}
