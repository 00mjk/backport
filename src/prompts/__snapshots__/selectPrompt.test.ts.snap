// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectPrompt should return a single answer 1`] = `
Array [
  Array [
    "? Pick ONE color … 
❯ black
  yellow
  red
  ─────",
  ],
  Array [
    "",
  ],
  Array [
    "? Pick ONE color … 
  black
❯ yellow
  red
  ─────",
  ],
  Array [
    "",
  ],
  Array [
    "✔ Pick ONE color · yellow as the sun",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
]
`;

exports[`selectPrompt should return multiple answers 1`] = `
Array [
  Array [
    "? Pick one or more colors … 
❯ ✔ black
  ✔ yellow
  ✔ red
    ─────",
  ],
  Array [
    "",
  ],
  Array [
    "? Pick one or more colors … 
❯ ✔ black
  ✔ yellow
  ✔ red
    ─────",
  ],
  Array [
    "",
  ],
  Array [
    "? Pick one or more colors … 
❯ ✔ black
  ✔ yellow
  ✔ red
    ─────",
  ],
  Array [
    "",
  ],
  Array [
    "? Pick one or more colors … 
  ✔ black
❯ ✔ yellow
  ✔ red
    ─────",
  ],
  Array [
    "",
  ],
  Array [
    "? Pick one or more colors … 
  ✔ black
❯ ✔ yellow
  ✔ red
    ─────",
  ],
  Array [
    "",
  ],
  Array [
    "✔ Pick one or more colors · b, y",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
]
`;
